File: 7-Grafos/2440-Familias-de-Troia/2440.py
def dfs(graph):
    visited, stack = set(), []
    counter = 0
    for node in graph:
        if node not in visited:
            counter += 1
            stack.append(node)
            while stack:
                v = stack.pop()
                if v not in visited:
                    visited.add(v)
                    stack.extend(reversed(graph[v]))
    return counter

nvertices, nedges = input().split()
nvertices, nedges = int(nvertices), int(nedges)
graph = {}
for i in range(1, nvertices + 1):
    graph.update({i : []})
for i in range (nedges):
    vertex, neighbor = input().split()
    vertex, neighbor = int(vertex), int(neighbor)
    graph[vertex].append(neighbor)
    graph[neighbor].append(vertex)
print(dfs(graph))

File: 7-Grafos/2440-Familias-de-Troia/2440.cpp
#include <iostream>
#include <vector>

void dfs(std::vector<bool> &visited, const std::vector<std::vector<int>> &graph, int node) {
    visited[node] = true;
    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) dfs(visited, graph, neighbor);
    }
}

int main() {
    int vertices, edges, u, v, counter = 0;
    std::cin>>vertices>>edges;
    std::vector<std::vector<int>> graph(vertices);
    std::vector<bool> visited(vertices, false);
    for (int i = 0; i < edges; ++i) {
        std::cin>>u>>v;
        --u; --v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    for (int i = 0; i < vertices; ++i) {
        if (!visited[i]) {
            ++counter;
            dfs(visited, graph, i);
        }
    }
    std::cout<<counter<<"\n";

    return 0;
}